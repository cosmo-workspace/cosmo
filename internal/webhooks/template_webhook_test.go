package webhooks

import (
	"context"
	"os"
	"time"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	"sigs.k8s.io/controller-runtime/pkg/client"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	cosmov1alpha1 "github.com/cosmo-workspace/cosmo/api/core/v1alpha1"
	wsv1alpha1 "github.com/cosmo-workspace/cosmo/api/workspace/v1alpha1"
	"github.com/cosmo-workspace/cosmo/pkg/kubeutil"
)

var _ = Describe("Template webhook", func() {
	Context("when creating template without urlbase", func() {
		It("should pass with defaulting urlbase", func() {
			ctx := context.Background()

			wsConfig := wsv1alpha1.Config{
				DeploymentName:      "ws-dep",
				ServiceName:         "ws-svc",
				IngressName:         "ws-ing",
				ServiceMainPortName: "mainPort",
				URLBase:             "https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{NAMESPACE}}.example.com",
			}

			tmpl := cosmov1alpha1.Template{
				ObjectMeta: metav1.ObjectMeta{
					Name: "code-server-test-wh",
					Labels: map[string]string{
						cosmov1alpha1.TemplateLabelKeyType: wsv1alpha1.TemplateTypeWorkspace,
					},
					Annotations: map[string]string{
						wsv1alpha1.TemplateAnnKeyWorkspaceDeployment:      wsConfig.DeploymentName,
						wsv1alpha1.TemplateAnnKeyWorkspaceIngress:         wsConfig.IngressName,
						wsv1alpha1.TemplateAnnKeyWorkspaceService:         wsConfig.ServiceName,
						wsv1alpha1.TemplateAnnKeyWorkspaceServiceMainPort: wsConfig.ServiceMainPortName,
						// no urlbase
						// wsv1alpha1.TemplateAnnKeyURLBase:                  wsConfig.URLBase,
					},
				},
				Spec: cosmov1alpha1.TemplateSpec{
					RawYaml: `apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  labels:
	cosmo/template: '{{INSTANCE}}'
	cosmo/template: code-server-test
  name: ws-ing
  namespace: '{{NAMESPACE}}'
spec:
  rules:
  - host: 'main-{{INSTANCE}}-{{NAMESPACE}}.{{DOMAIN}}'
	http:
	  paths:
	  - path: /
		pathType: Prefix
		backend:
		  service:
			name: ws-svc
			port: 
			  number: 8080
---
apiVersion: v1
kind: Service
metadata:
  labels:
	cosmo/template: '{{INSTANCE}}'
	cosmo/template: code-server-test
  name: ws-svc
  namespace: '{{NAMESPACE}}'
spec:
  ports:
  - name: main
	port: 8080
	protocol: TCP
  selector:
	cosmo/template: '{{INSTANCE}}'
	cosmo/template: code-server-test
  type: ClusterIP
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
	cosmo/template: '{{INSTANCE}}'
	cosmo/template: code-server-test
  name: ws-dep
  namespace: '{{NAMESPACE}}'
spec:
  replicas: 1
  selector:
	matchLabels:
	  cosmo/template: '{{INSTANCE}}'
	  cosmo/template: code-server-test
  template:
	metadata:
	  labels:
		cosmo/template: '{{INSTANCE}}'
		cosmo/template: code-server-test
	spec:
	  containers:
	  - image: 'code-server:{{IMAGE_TAG}}'
		name: code-server-test
		ports:
		- containerPort: 8080
		  name: main
		  protocol: TCP
`,
					RequiredVars: []cosmov1alpha1.RequiredVarSpec{
						{
							Var: "{{DOMAIN}}",
						},
						{
							Var:     "{{IMAGE_TAG}}",
							Default: "latest",
						},
					},
				},
			}

			err := k8sClient.Create(ctx, &tmpl)
			Expect(err).ShouldNot(HaveOccurred())

			expectedTmpl := tmpl.DeepCopy()
			expectedTmpl.TypeMeta = metav1.TypeMeta{
				Kind:       "Template",
				APIVersion: cosmov1alpha1.GroupVersion.String(),
			}
			expectedTmpl.Annotations[wsv1alpha1.TemplateAnnKeyURLBase] = DefaultURLBase

			var createdTmpl cosmov1alpha1.Template
			Eventually(func() error {
				err := k8sClient.Get(ctx, client.ObjectKey{Name: tmpl.GetName(), Namespace: tmpl.GetNamespace()}, &createdTmpl)
				if err != nil {
					return err
				}
				return nil
			}, time.Second*10).Should(Succeed())

			eq := kubeutil.LooseDeepEqual(&createdTmpl, expectedTmpl, kubeutil.WithPrintDiff(os.Stderr))
			Expect(eq).Should(BeTrue())
		})
	})
})
