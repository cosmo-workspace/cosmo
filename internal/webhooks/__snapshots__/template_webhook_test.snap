['Template webhook when creating template without urlbase should pass with defaulting urlbase 1']
SnapShot = """
{
  \"kind\": \"Template\",
  \"apiVersion\": \"cosmo-workspace.github.io/v1alpha1\",
  \"metadata\": {
    \"name\": \"code-server-test-wh\",
    \"creationTimestamp\": null,
    \"labels\": {
      \"cosmo-workspace.github.io/type\": \"workspace\"
    },
    \"annotations\": {
      \"workspace.cosmo-workspace.github.io/deployment\": \"ws-dep\",
      \"workspace.cosmo-workspace.github.io/service\": \"ws-svc\",
      \"workspace.cosmo-workspace.github.io/service-main-port\": \"mainPort\",
      \"workspace.cosmo-workspace.github.io/urlbase\": \"https://{{NETRULE_GROUP}}-{{INSTANCE}}-{{USER_NAME}}.domain\"
    }
  },
  \"spec\": {
    \"requiredVars\": [
      {
        \"var\": \"{{DOMAIN}}\"
      },
      {
        \"var\": \"{{IMAGE_TAG}}\",
        \"default\": \"latest\"
      }
    ],
    \"rawYaml\": \"apiVersion: networking.k8s.io/v1\\nkind: Ingress\\nmetadata:\\n  labels:\\n    cosmo-workspace.github.io/template: '{{INSTANCE}}'\\n    cosmo-workspace.github.io/template: code-server-test\\n  name: ws-ing\\n  namespace: '{{NAMESPACE}}'\\nspec:\\n  rules:\\n  - host: 'main-{{INSTANCE}}-{{NAMESPACE}}.{{DOMAIN}}'\\n    http:\\n      paths:\\n      - path: /\\n        pathType: Prefix\\n        backend:\\n          service:\\n            name: ws-svc\\n            port: \\n              number: 8080\\n---\\napiVersion: v1\\nkind: Service\\nmetadata:\\n  labels:\\n    cosmo-workspace.github.io/template: '{{INSTANCE}}'\\n    cosmo-workspace.github.io/template: code-server-test\\n  name: ws-svc\\n  namespace: '{{NAMESPACE}}'\\nspec:\\n  ports:\\n  - name: main\\n    port: 8080\\n    protocol: TCP\\n  selector:\\n    cosmo-workspace.github.io/template: '{{INSTANCE}}'\\n    cosmo-workspace.github.io/template: code-server-test\\n  type: ClusterIP\\n---\\napiVersion: apps/v1\\nkind: Deployment\\nmetadata:\\n  labels:\\n    cosmo-workspace.github.io/template: '{{INSTANCE}}'\\n    cosmo-workspace.github.io/template: code-server-test\\n  name: ws-dep\\n  namespace: '{{NAMESPACE}}'\\nspec:\\n  replicas: 1\\n  selector:\\n    matchLabels:\\n      cosmo-workspace.github.io/template: '{{INSTANCE}}'\\n      cosmo-workspace.github.io/template: code-server-test\\n  template:\\n    metadata:\\n      labels:\\n        cosmo-workspace.github.io/template: '{{INSTANCE}}'\\n        cosmo-workspace.github.io/template: code-server-test\\n    spec:\\n      containers:\\n      - image: 'code-server:{{IMAGE_TAG}}'\\n        name: code-server-test\\n        ports:\\n        - containerPort: 8080\\n          name: main\\n          protocol: TCP\\n\"
  }
}
"""

['Template webhook when including ClusterRole in Template should pass with warning even though invalid scope 1']
SnapShot = """
{
  \"kind\": \"Template\",
  \"apiVersion\": \"cosmo-workspace.github.io/v1alpha1\",
  \"metadata\": {
    \"name\": \"cluster-level-tmpl\",
    \"creationTimestamp\": null
  },
  \"spec\": {
    \"rawYaml\": \"apiVersion: rbac.authorization.k8s.io/v1\\nkind: ClusterRole\\nmetadata:\\n  name: privileged\\n  namespace: {{NAMESPACE}}\\nrules:\\n- apiGroups:\\n  - '*'\\n  resources:\\n  - '*'\\n  verbs:\\n  - '*'\\n- nonResourceURLs:\\n  - '*'\\n  verbs:\\n  - '*'\\n\"
  }
}
"""

['Template webhook when including ClusterRole in Template with skip validation annotation should pass with warning even though invalid scope 1']
SnapShot = """
{
  \"kind\": \"Template\",
  \"apiVersion\": \"cosmo-workspace.github.io/v1alpha1\",
  \"metadata\": {
    \"name\": \"cluster-level-tmpl-passed\",
    \"creationTimestamp\": null,
    \"annotations\": {
      \"cosmo-workspace.github.io/skip-validation\": \"1\"
    }
  },
  \"spec\": {
    \"rawYaml\": \"apiVersion: rbac.authorization.k8s.io/v1\\nkind: ClusterRole\\nmetadata:\\n  name: privileged\\n  namespace: {{NAMESPACE}}\\nrules:\\n- apiGroups:\\n  - '*'\\n  resources:\\n  - '*'\\n  verbs:\\n  - '*'\\n- nonResourceURLs:\\n  - '*'\\n  verbs:\\n  - '*'\\n\"
  }
}
"""

['Template webhook when including Pod in ClusterTemplate should pass with warning even though invalid scope 1']
SnapShot = """
{
  \"kind\": \"ClusterTemplate\",
  \"apiVersion\": \"cosmo-workspace.github.io/v1alpha1\",
  \"metadata\": {
    \"name\": \"ns-level-ctmpl\",
    \"creationTimestamp\": null
  },
  \"spec\": {
    \"rawYaml\": \"apiVersion: v1\\nkind: Pod\\nmetadata:\\n  name: nginx\\n  namespace: {{NAMESPACE}}\\nspec:\\n  containers:\\n  - name: nginx\\n    image: nginx:alpine\\n\"
  }
}
"""
